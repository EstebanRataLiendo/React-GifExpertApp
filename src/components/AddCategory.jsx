/*
Componente que permite tomar el Enter y retornar un formulario con un input
Importamos el useState.
Exportamos la app y desestructuramos la propiedad "onNewCategory" de la app principal GifExpertApp
Usamos useState para retornar un valor y actualizarlo.
-onInputChange (la constante será igual al resultado de una función_
la cual toma cada cambio del setInputValue y la actualiza luego en el input [onChange])
-onSubmit (cuando presiono enter se ejecuta una función que actualiza el formulario)
-preventDefault (evita que se recargue la pagina en cada cambio)
Si el input esta vacío o con un solo caracter, no se ejecuta la función
Le da a la propiedad onNewCategory el valor del input sin espacios al comienzo y al final
Volvemos a poner en blanco el input.
Retornamos lo que vamos a renderizar:
    -form que se actualiza cuando presionamos enter
    -input con propiedades: tipo, placeholder, valor igual al input y 
    manejador de evento onChange que ejecuta onInputChange (sin esto no nos deja escribir en el input)
*/
import { useState } from "react"

export const AddCategory = ({onNewCategory}) => {

    const [inputValue, setInputValue] = useState('');

    const onInputChange = (event) => {
        setInputValue(event.target.value)
    }
    const onSubmit= (e) => {
        e.preventDefault();
        if (inputValue.trim().length<=1) return;
        onNewCategory(inputValue.trim())
        setInputValue('');
    }
    return (
        <>
            <form onSubmit={onSubmit}>
                <input
                    type="text"
                    placeholder="Buscar gifs"
                    value={inputValue}
                    onChange={onInputChange}
                />
            </form>
        </>
            )
}

