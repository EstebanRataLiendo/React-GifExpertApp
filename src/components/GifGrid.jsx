/* 
Importamos el componente GifItem (Devuelve el url y titulo de los gif)
Importamos el hook useFetchGifs
Exportamos el componente GifGrid
Usamos el hook que creamos para tomar las imagenes y el Loading
Retornamos un <h3> con las category's
Preguntamos si isLoading es true, renderizamos <h2>
Retornamos un <div> con formato de css y mapeamos las imagenes tomadas del hook
En el mapeo usamos el componente GifItem para traer y renderizar las propiedades url, id y title
*/

import { GifItem } from "./";
import { useFetchGifs } from "../hooks/useFetchGifs";

export const GifGrid = ({category}) => {
    const {images, isLoading} = useFetchGifs(category);

    return (
        <>
            <h3>{category} </h3>
            {
                isLoading && (<h2>Cargando...</h2>)
            }

            <div className="card-grid">
            {
                images.map((image) =>(
                        <GifItem 
                            key={image.id}
                            {...image} //exparse todas las propiedades (url, id y title)
                        />
                ))
            }
            </div>
        </>
    )
}