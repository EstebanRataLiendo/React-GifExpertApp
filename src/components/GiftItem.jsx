//componente que devuelve los gifs con sus titulos
//Exportamos la funciÃ³n al GifGrid
export const GifItem = ({title, url}) => {
    return ( 
        <div className="card">
            <img src={url} alt={title} />
            <p>{title}</p>
        </div>
    );
}