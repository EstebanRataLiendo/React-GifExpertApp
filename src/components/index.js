//Aqui agrupamos nuestros componentes

export * from './AddCategory'
export * from './GifGrid'
export * from './GifItem'