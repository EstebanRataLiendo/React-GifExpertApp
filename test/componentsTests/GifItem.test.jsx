import {  fireEvent, render,screen } from "@testing-library/react"
import { GifItem } from "../../src/components/GifItem";

const title = 'Prueba'
const url  = 'http://prueba-rata.com/'
describe('GifItem', () => {
    test('Debe hacer match con el snapshot ', () => {
        const {container}= render(<GifItem title={title} url={url} />);
        expect (container).toMatchSnapshot();
    });

    test('debe mostrar la imagen con el url y alt indicado', () => {
        render(<GifItem title={title} url={url} />);
        // screen.debug()
        // console.log(screen.getByRole('img').src)

        expect(screen.getByRole('img').src).toBe(url);
        expect(screen.getByRole('img').alt).toBe(title);
    })

    test('debe mostrar el titulo en el componente', () => {
        render(<GifItem title={title} url={url} />);
        expect(screen.getByText(title)).toBeTruthy();
    })
    
})
