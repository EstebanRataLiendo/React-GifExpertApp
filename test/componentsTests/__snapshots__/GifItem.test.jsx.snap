// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GifItem Debe hacer match con el snapshot  1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Prueba"
      src="http://prueba-rata.com/"
    />
    <p>
      Prueba
    </p>
  </div>
</div>
`;
